src/main/java/carpet/mixins/PistonBlock_movableTEMixin.java

```diff
@@ -0,0 +1,123 @@
+            Direction direction_1, boolean boolean_1, Direction direction_2, CallbackInfoReturnable<Boolean> cir)
+    {
+        Block block_1 = blockState_1.getBlock();
+        //Make CommandBlocks movable, either use instanceof CommandBlock or the 3 cmd block objects,
+        if (CarpetSettings.b_movableTileEntities && block_1 instanceof CommandBlock)
+        {
+            cir.setReturnValue(true);
+        }
+    }
+    
+    private static boolean isPushableTileEntityBlock(Block block)
+    {
+        //Making PISTON_EXTENSION (BlockPistonMoving) pushable would not work as its createNewTileEntity()-method returns null
+        return block != Blocks.ENDER_CHEST && block != Blocks.ENCHANTING_TABLE &&
+                       block != Blocks.END_GATEWAY && block != Blocks.END_PORTAL && block != Blocks.MOVING_PISTON  &&
+                       block != Blocks.SPAWNER;
+    }
+    
+    @Redirect(method = "isMovable", at = @At(value = "INVOKE", target = "Lnet/minecraft/block/Block;hasBlockEntity()Z"))
+    private static boolean ifHasBlockEntity(Block block)
+    {
+        if (!block.hasBlockEntity())
+        {
+            return !true;
+        }
+        else
+        {
+            return !(CarpetSettings.b_movableTileEntities && isPushableTileEntityBlock(block));
+        }
+    }
+    @Inject(method = "move", at = @At(value = "INVOKE", shift = At.Shift.BEFORE,
+            target = "Ljava/util/List;size()I", ordinal = 4),locals = LocalCapture.CAPTURE_FAILHARD)
+    private void onMove(World world_1, BlockPos blockPos_1, Direction direction_1, boolean boolean_1,
+            CallbackInfoReturnable<Boolean> cir, BlockPos blockPos_2, PistonHandler pistonHandler_1,
+            List<BlockPos> list_1, List<BlockState> list_2, List list_3, int int_2, BlockState blockStates_1[],
+            Direction direction_2, Set set_1)
+    {
+        //Get the blockEntities and remove them from the world before any magic starts to happen
+        if (CarpetSettings.b_movableTileEntities)
+        {
+            list1_BlockEntities.set(Lists.<BlockEntity>newArrayList());
+            for (int i = 0; i < list_1.size(); ++i)
+            {
+                BlockPos blockpos = list_1.get(i);
+                BlockEntity blockEntity = (list_2.get(i).getBlock().hasBlockEntity()) ? world_1.getBlockEntity(blockpos) : null;
+                list1_BlockEntities.get().add(blockEntity);
+                if (blockEntity != null)
+                {
+                    //hopefully this call won't have any side effects in the future, such as dropping all the BlockEntity's items
+                    //we want to place this same(!) BlockEntity object into the world later when the movement stops again
+                    world_1.removeBlockEntity(blockpos);
+                    blockEntity.markDirty();
+                }
+            }
+        }
+    }
+    
+    @Inject(method = "move", at = @At(value = "INVOKE", shift = At.Shift.BEFORE,
```

Comment:
why this is inject, not redirect. You are essentially doubling on BE's, one is vanilla and the second is this one. The vanilla setBlockEntity will be called right after right?

Reply:
You can't capture local variables with @Redirects..the very next mixin cancels the same call